Conclusion
----------
The goal of this bachelor internship was to first and foremost be able to use python for data analysis and data visualization. Secondly and more importantly, this reports objective was to analyze ctDNA derived from Cyclomics and set the foundation for the construction of a data filter. In order to achieve this goal, analyzing was conducted and separated between fastq and VCF files.

Fastq files contain regions with high and low mean quality score. Identification of these regions could help us understand what chunks of the sequence are difficult for the nanopore to sequence. Regions that have an average high score in between runs could then be identified as method specific high quality score regions. These regions have a higher probability of containing an error and are thus less suitable for variant calling. Therefore, our data filter should exclude these sequences. Likewise, regions that have low quality scores have less probability to contain an error and should thus be included in data filtering for variant calling. However, while in theory quality score analysis sounds promising, analyzing of different datasets and even chunks of a same datasets divided into regions showed the inconsistency in these quality scores. The datasets mean quality score differs heavily between these datasets and further clustering showed little to no correlation between specific regions and their quality score. Nevertheless, we have not established enough data to be able to dismiss quality score analysis entirely for the construction of a data filter and suggest that further investigations should be conducted on comparing regions between chunks of the datasets located in the similar clusters. Hopefully, these analyses would show that although the average quality score differs heavily between chunks and datasets, still same regions would be located in the same cluster groups for each chunk.

Analysing SNPs in VCF files was conducted to both identify real driver mutations located in a p53 mutated database and visualize other mutations already present in p53 wild-type. These other mutations consist of real passenger mutations, but also PCR and sequencing errors. These alterations identified in wild-type could be applied for data filtering. In our research, we were able to identify real driver mutations in predetermined p53 mutant datasets. Also, regions could be identified harboring both real passenger mutations and errors. The specific position of the passenger mutation can be included in a data filter, removing this passenger mutations background noise. Furthermore, error-prone regions can also be included in a data filter, removing background noise anywhere these regions persists in the sequence. However, the current lack of automation on this process of identifying passenger mutations and error-prone regions make the construction of a data filter difficult. Therefore, future investigation should be conducted on the development of automation for this process. Additionally, after completion, multiple datasets covering both p53 and other known tumor driver genes should be investigated. Analyzing other p53 datasets could provide to empower described error-prone regions, while datasets covering other tumor driver genes could analyze these regions for application in other datasets. Possible, the error-prone regions are very similar between genes, simplifying data filter construction. In contrast, error-prone regions could vary heavily between genes. If a data filter can be devised for a p53 database, other frequently mutated genes should be analyzed using this filter, aiming either to expanding the data filter to apply at a broader range or construct data filters for individual genes using the same methods shown here depending on the similarities of error-prone regions between genes.

In summary, this report was able to investigate ctDNA from Cyclomics to acquire a better understanding into creating a data filter. In the future, more Cyclomics databases should be constructed and analyzed using covered scripts. This could both help identifying whether there is a correlation between regions and quality scores for fastq files and whether SNP exclusion can be applied to data filtering for VCF files.

Personal reflection
+++++++++++++++++++
For my internship, I wanted to try something I have not yet experienced in my bachelor and because I have always been interested in programming, I ought to try python. Learning and understanding a whole new language was both challenging but all the more rewarding. I am very glad to be able to say, that after this internship, I am now able to write and understand python. Furthermore, I could have never imagined having achieved this amount of knowledge of programming and a whole library_ containing only self-written functions and scripts. Also, during this internship, I experienced some minor setbacks with data that was not in line with our expectations and while this is unfortunate, it also gave me a more realistic view upon biomedical research. Altogether, this internship gave me a really positive and realistic view on research and encouraged me to pursue a research master to further investigate my interest in biomedical research. 

Acknowledgements
++++++++++++++++
I would like to thank Alessio Marcozzi for his guidance and support during this project. In addition, I would like to thank Dr. Wigard Kloosterman for giving me this wonderful opportunity and for his supervision during my internship at the Department of Genetics.

.. _library: https://github.com/DouweSpaanderman/NaDA/

|
|

.. figure:: https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Next.png
   :align:  center
   :width: 100px
   :height: 100px
   :target: https://rawgit.com/DouweSpaanderman/NaDA/blob/master/Documentation/build/html/References.html