
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Conclusion &#8212; Internship report 3.0a1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '3.0a1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="References" href="References.html" />
    <link rel="prev" title="Results and discussion" href="Results and Discussion.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="conclusion">
<h1>Conclusion<a class="headerlink" href="#conclusion" title="Permalink to this headline">¶</a></h1>
<p>The goal of this bachelor internship was to first and foremost be able to use python for data analysis and data visualization. Secondly and more importantly, this report’s objective was to analyse ctDNA derived from Cyclomics and set the foundation for construction of a data filter. In order to achieve this goal, analysing was conducted and seperated between fastq and vcf files.</p>
<p>Fastq files contain regions with high and low mean quality score. Identification of these regions could help us understand what chunks of the sequence are difficult for the nanopore to sequence. Chunks that have an average high score in between runs could then be identified as method specific high quality score chunks. These chunks have a higher probability of containing an error and are thus less suited for variant calling. Therefor, our data filter should exclude these sequences. Likewise, chunks that have low quality scores have less probability to contain an error and should thus be included in data filtering for variant calling. However, while in theory quality score analysis sounds promising, analysing of different datasets and even same datasets divided in chunks showed the inconsistency in these quality scores. The datasets mean quality score differs heavily between these datasets and further clustering showed little to no correlation between specific regions and their quality score. Nevertheless, we haven’t established enough data to be able to dismiss quality score analysis entirely for the construction of a data filter and suggest that further investigations should be conducted on comparing chunks between datasets located in the highest score cluster. Hopefully, these analyses would show that although the average quality score differs heavily between datasets, still same chunks would be located in high score clusters.</p>
<p>Analysing SNPs in vcf files was conducted to both identify real mutations located in a p53 mutated database and visualize other mutations already present in p53 wild-type. Other mutations consist of real passenger mutations, but also pcr and sequencing errors. These alterations identified in wild-type could be applied for data filtering. Currently, we have yet to establish enough data to identify method specific background noise. However, further Cyclomics database analyses should be quick as scripts have already been prepared and optimised for new analyses. With this in mind, constructing more p53 databases with Cyclomics should have the priority to be able to identify if same regions as shown in this report will contain SNPs. If so data filter could be constructed for p53 otherwise we would advise to construct a data filter using cfDNA for every patient. Also, if a data filter can be devised for p53 database, other frequently mutated genes should be analysed using this filter, either expending the data filter to apply at a broader range or construct data filters for individual genes using the same methods shown here.</p>
<p>In summary, this report was able to investigate ctDNA from Cyclomics to acquire a better understanding to creating a data filter. Currently, more Cyclomics databases should be constructed and analysed using covered scripts. This could both help identify if there is a correlation between regions and quality scores for fastq files and if SNP exclusion can be applied to data filtering for vcf files.</p>
<div class="section" id="personal-reflection">
<h2>Personal reflection<a class="headerlink" href="#personal-reflection" title="Permalink to this headline">¶</a></h2>
<p>For my internship, I wanted to try something I haven’t yet experienced in my bachelor and because I have always been interested in programming, I ought to try python. Learning and understanding a whole new language was both challenging but all the more rewarding. I’m very glad to be able to say, that after this internship, I’m now able to write and understand python. Furthermore, I could have never imagined to have achieved this amount of knowledge of programming and a whole <a class="reference external" href="https://github.com/DouweSpaanderman/NaDA/">library</a> containing only self written functions and scripts. Also, during this internship, I experienced some minor setbacks with data that was not in line with our expectations and while this is unfortunately, it also gave me a more realistic view upon biomedical research. All in all this internship gave me a really positive and realistic view on research and encouraged me to pursue a research master to further investigate my interest in biomedical research.</p>
</div>
<div class="section" id="acknowledgements">
<h2>Acknowledgements<a class="headerlink" href="#acknowledgements" title="Permalink to this headline">¶</a></h2>
<p>I would like to thank Alessio Marcozzi for his guidance and support during this project. In addition, I would like to thank Dr. Wigard Kloosterman for giving me this wonderful opportunity and for his supervision during my internship at the department of Genetics.</p>
<div class="line-block">
<div class="line"><br /></div>
<div class="line"><br /></div>
</div>
<div class="figure align-center">
<a class="reference external image-reference" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/build/html/References.html"><img alt="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Next.png" src="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Next.png" style="width: 100px; height: 100px;" /></a>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">Internship report</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="General concept.html">General concept</a></li>
<li class="toctree-l1"><a class="reference internal" href="Results and Discussion.html">Results and discussion</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Conclusion</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#personal-reflection">Personal reflection</a></li>
<li class="toctree-l2"><a class="reference internal" href="#acknowledgements">Acknowledgements</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="References.html">References</a></li>
<li class="toctree-l1"><a class="reference internal" href="Supplementary.html">Supplementary</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="Results and Discussion.html" title="previous chapter">Results and discussion</a></li>
      <li>Next: <a href="References.html" title="next chapter">References</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, Douwe Spaanderman.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/Conclusion.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>