
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Results and discussion &#8212; Internship report 3.0a1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '3.0a1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Conclusion" href="Conclusion.html" />
    <link rel="prev" title="General concept" href="General concept.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="results-and-discussion">
<h1>Results and discussion<a class="headerlink" href="#results-and-discussion" title="Permalink to this headline">¶</a></h1>
<div class="section" id="fastq-files">
<h2>Fastq files<a class="headerlink" href="#fastq-files" title="Permalink to this headline">¶</a></h2>
<p><strong>Quality score distribution</strong>. Fastq is a text-based format derived from sequencers such as Illumina and Nanopore that contain a biological sequence and for every base a quality score.
this quality score represents the sequencers ability to call a base as true, in other words, the score gives an indication about how sure the sequencer is that the output it gives is the base that is actually present in the sample. If the score for the base is low, it’s most likely the right base, while if the score is high the machine is unsure if it’s the real base. Nanopore sequencer measures changes in current to identify bases in the sample and if changes in this current are less clear a high quality score is given for this base.</p>
<p>Analyzing quality scores on sequences is important to be able to understand which regions of bases are difficult to get base called. More importantly, a mutation called in a region with low quality score is more likely to be a real mutation in contrast to a mutation that occurs through wrong base calling in areas with high quality scores. Clarifying which regions have high scores is important for achieving tests with high specificity, as these regions will not be selected for variant calling. Likewise, regions with a low score are good targets for identifying mutations and are thus able to increase test sensitivity. Therefore the aim of fastq file analysis is to identify regions of bases with high and low scores, as these regions can be used for data filtering. In data filtering, high score regions can be excluded and low score regions can be included because mutations identified in high score regions are more likely to be sequencing errors.</p>
<p>In order to analyze high and low score regions, fastq files first need to be stripped of bases and their scores. the following function was written to strip the files of data:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">parse_fasta_file_error</span><span class="p">(</span><span class="n">sequence_file</span><span class="p">):</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">sequence_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">id_</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="n">sequence</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">]:</span>
                                <span class="k">continue</span>
                        <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;@&#39;</span><span class="p">):</span>
                                <span class="k">if</span> <span class="ow">not</span> <span class="n">id_</span><span class="p">:</span> <span class="c1">#if id_ == False</span>
                                        <span class="n">id_</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                                <span class="k">else</span><span class="p">:</span>
                                        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;WARNING: An ID was found</span>
                                                <span class="n">without</span> <span class="n">corresponding</span>
                                                <span class="n">sequence</span><span class="o">.</span><span class="s1">&#39;, id_)</span>
                                        <span class="n">id_</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                        <span class="k">elif</span> <span class="n">id_</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">sequence</span><span class="p">:</span>
                                <span class="n">sequence</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                        <span class="k">elif</span> <span class="n">id_</span> <span class="ow">and</span> <span class="n">sequence</span><span class="p">:</span>
                                <span class="n">score</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                                <span class="n">data</span><span class="p">[</span><span class="n">id_</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;sequence&#39;</span><span class="p">:</span><span class="n">sequence</span><span class="p">,</span>
                                                <span class="s1">&#39;score&#39;</span><span class="p">:</span><span class="n">score</span><span class="p">}</span>
                                <span class="n">sequence</span> <span class="o">=</span> <span class="kc">False</span>
                                <span class="n">id_</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">return</span> <span class="n">data</span>
</pre></div>
</div>
<p>The function returns a dictionary with identification (id), sequence and scores for individual bases in the fastq file. With the data stripped from the fastq file, the first question was whether there are changes in quality score present in our file. Therefore, the distribution of the error rate was visualized for every individual base in the sequence as shown in Figure 2, which illustrates the occurrence of high and low score regions. In these high score regions, bases have more chance to be falsely assigned and are thus less reliable for mutation identification. Likewise, mutations found in low score regions are more likely to be rightly assigned and thus true mutations. Figure 2 clearly visualizes changes in quality score distribution and could thus indicate that high and low regions are present in our sample.</p>
<div class="figure align-center" id="id4">
<a class="reference internal image-reference" href="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Fastq_files_qualityscore.png"><img alt="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Fastq_files_qualityscore.png" src="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Fastq_files_qualityscore.png" style="width: 700px; height: 450px;" /></a>
<p class="caption"><span class="caption-text">Figure 2: <strong>Distribution of error rate (quality score) for every individual base.</strong> In this Figure, every 10th base is visualized on the X-axis to clarify the data. The mean of all the quality scores is 0.21.</span></p>
</div>
<p>While Figure 2 gives a good overview on the importance of analyzing quality scores, one fastq file has little quantifiable value for identifying high score regions. To further investigate our first question, whether changes in quality scores are present in our samples, 1139 fastq files from Nanopore with reads on chromosome 9 were likewise investigated for quality score distribution. Firstly, mean scores of the quality scores were plotted in Figure 3. The mean scores of the files are a good first indication of the sequence quality and can be used to filter and select files for variant calling, in which a low mean quality score is important. Similar to Figure 2, mean score distribution also visualizes changes in quality scores between files further empowering the possibility that high and low regions are present in our samples.</p>
<div class="figure align-center" id="id5">
<a class="reference internal image-reference" href="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Mean_distribution_of_all_quality_score.png"><img alt="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Mean_distribution_of_all_quality_score.png" src="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Mean_distribution_of_all_quality_score.png" style="width: 700px; height: 450px;" /></a>
<p class="caption"><span class="caption-text">Figure 3: <strong>Mean distribution of quality scores</strong>. Mean scores from 1139 fastq files were calculated and visualised in a violin plot.</span></p>
</div>
<p><strong>Region quality score distribution</strong>. Now we know that quality score changes occur in our samples, the next question arose whether these changes translate into regions with high and low scores. In order to investigate if these regions are present in our sample, regions of bases were selected instead of single bases. Also, these regions were investigated in different sizes to analyze the impact of different sized regions on the occurrence of high and low score region. Therefore, sequences were chunked into pieces of four (tetramer) to six (hexamer) bases and for every sequence chunk, its respected mean score was calculated. Size and overlap of the chunks of sequences could be selected by the following function:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">split_overlap</span><span class="p">(</span><span class="n">iterable</span><span class="p">,</span><span class="n">size</span><span class="p">,</span><span class="n">overlap</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">size</span> <span class="o">&lt;</span> <span class="mi">1</span> <span class="ow">or</span> <span class="n">overlap</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;&quot;size&quot; must be an integer with</span>
                                 <span class="o">&gt;=</span> <span class="mi">1</span> <span class="k">while</span> <span class="s2">&quot;overlap&quot;</span> <span class="n">must</span> <span class="n">be</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="s1">&#39;)</span>
        <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">iterable</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">size</span><span class="p">:</span>
                        <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">iterable</span><span class="p">)</span>
                        <span class="k">return</span> <span class="n">result</span>
                <span class="k">else</span><span class="p">:</span>
                        <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">iterable</span><span class="p">[:</span><span class="n">size</span><span class="p">])</span>
                        <span class="n">iterable</span> <span class="o">=</span> <span class="n">iterable</span><span class="p">[</span><span class="n">size</span><span class="o">-</span><span class="n">overlap</span><span class="p">:]</span>
</pre></div>
</div>
<p>Following, the mean of all the quality scores of the same chunks of sequences were either plotted directly (Figure 4 A - C) or indirectly after being divided into categories of high, medium and low quality score (Figure 4 D - F). Categorizing was performed before calculating the mean of a sequence, aiming to manipulate and increase data analysis. Subsequently, categories were counted for each sequence. Parameters for categorizing were randomly selected and differentiated for each size, because with larger regions the mean of the quality score shift further towards medium. To account for this difference the high requirements are lowered and the low requirements were increased as follows:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">high_medium_low_scores</span><span class="p">(</span><span class="n">listed_scores</span><span class="p">,</span> <span class="n">size</span><span class="p">):</span>
        <span class="n">group_score</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">listed_scores</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">s</span> <span class="o">&gt;=</span> <span class="p">(</span><span class="mf">0.40</span><span class="o">-</span><span class="mf">0.02</span><span class="o">*</span><span class="n">size</span><span class="p">):</span>
                        <span class="n">group_score</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;High&#39;</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">s</span> <span class="o">&lt;=</span> <span class="p">(</span><span class="mf">0.15</span><span class="o">+</span><span class="mf">0.01</span><span class="o">*</span><span class="n">size</span><span class="p">):</span>
                        <span class="n">group_score</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Low&#39;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                        <span class="n">group_score</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Medium&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">group_score</span>
</pre></div>
</div>
<div class="figure align-center" id="id6">
<a class="reference internal image-reference" href="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Fastq_gridplots.png"><img alt="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Fastq_gridplots.png" src="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Fastq_gridplots.png" style="width: 700px; height: 450px;" /></a>
<p class="caption"><span class="caption-text">Figure 4: <strong>Quality score analysis in 6 scenarios.</strong> A - C) Mean score for all combinations in chunk sizes (A = 4, B = 5, C = 6) for 1139 fastq files derived from nanopore sequencing of chromosome 9. D - F) Scores for regions have been categorized into high, medium and low for regions of the same size as A to C. Next, the number of times a region was called under a certain category was counted and collected for the same data set. In these figures, scores are set in percentage of the total amount of times a region occurs in the data set.(Interactive figure at <a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/source/_static/gridplot.html">GridPlot</a>)</span></p>
</div>
<p>Table 1 highlights the highest and lowest five scoring sequences measured in Figure 4. The highest scoring sequence has the biggest chance to have wrongly assigned bases in its sequence. In contrast, bases in lower scoring sequences have been most likely rightfully assigned and are therefore indeed the right base. These findings should be taken into account when investigating mutations, as a mutation found in, for instance, TTCC is more likely to be a real mutation than an alteration found in GCTT. Conclusively, this dataset of chromosome 9 does indeed have high and low quality score regions which should be further analyzed for respectively exclusion and inclusion into a data filter.</p>
<table border="1" class="docutils">
<colgroup>
<col width="20%" />
<col width="13%" />
<col width="20%" />
<col width="13%" />
<col width="20%" />
<col width="13%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">C</th>
<th class="head">Score</th>
<th class="head">B</th>
<th class="head">Score</th>
<th class="head">C</th>
<th class="head">Score</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>GCTT</td>
<td>0.364</td>
<td>AGCTT</td>
<td>0.422</td>
<td>AGCTTT</td>
<td>0.501</td>
</tr>
<tr class="row-odd"><td>CTTG</td>
<td>0.353</td>
<td>GCCTT</td>
<td>0.405</td>
<td>TTCGCA</td>
<td>0.499</td>
</tr>
<tr class="row-even"><td>TAAT</td>
<td>0.313</td>
<td>GCTTG</td>
<td>0.393</td>
<td>GGGACG</td>
<td>0.489</td>
</tr>
<tr class="row-odd"><td>GTAG</td>
<td>0.298</td>
<td>GCTTA</td>
<td>0.372</td>
<td>CCATGT</td>
<td>0.482</td>
</tr>
<tr class="row-even"><td>TAGC</td>
<td>0.293</td>
<td>ATTGA</td>
<td>0.367</td>
<td>GAATCT</td>
<td>0.466</td>
</tr>
<tr class="row-odd"><td>…</td>
<td>&#160;</td>
<td>…</td>
<td>&#160;</td>
<td>…</td>
<td>&#160;</td>
</tr>
<tr class="row-even"><td>GGAT</td>
<td>0.136</td>
<td>TTAAA</td>
<td>0.112</td>
<td>CCTAAT</td>
<td>0.058</td>
</tr>
<tr class="row-odd"><td>CCCT</td>
<td>0.135</td>
<td>GTCTT</td>
<td>0.104</td>
<td>TTCACA</td>
<td>0.054</td>
</tr>
<tr class="row-even"><td>GTTC</td>
<td>0.131</td>
<td>TTGGA</td>
<td>0.100</td>
<td>TTTTTC</td>
<td>0.053</td>
</tr>
<tr class="row-odd"><td>CCTC</td>
<td>0.129</td>
<td>GGACC</td>
<td>0.098</td>
<td>CCAATC</td>
<td>0.050</td>
</tr>
<tr class="row-even"><td>TTCC</td>
<td>0.128</td>
<td>TTTTT</td>
<td>0.085</td>
<td>GGACGT</td>
<td>0.049</td>
</tr>
</tbody>
</table>
<blockquote>
<div>Table 1: <strong>Highest and lowest five scoring sequences</strong>. A - C) score is mean score for all combinations of same size and data set as figure 3.</div></blockquote>
<p><strong>Clustering</strong>. Overall, the fastq analysis showed the occurrence of high and low regions in our dataset. Next, these regions should be further investigated and clearer visualized for construction of our data filter. In order to analyze which regions should be included and excluded in our data filter clustering is conducted. Clustering is a method in which data points get coupled in groups (clusters) by a certain geometry. Here, K-Means is used for clustering, which makes clusters based on the distance between data points and a selected centroid. The centroid is the mean of a cluster and is defined by a trial and error process. This process is repeated until centroids are selected, which happens when the within-cluster sum of squares is minimized.</p>
<p>In Figure 5 three clusters are formed in which the yellow one represents sequences with often reported high score and few times reported low score. In this cluster, alterations are more likely to be falsely assigned. Furthermore, in the blue cluster, alterations are more likely to be rightfully assigned, cause sequences are often reported in low score and only a few times in high score. A great advantage of clustering of data is that a clearer view on which sequences to include and exclude for mutation calling can be provided. Here, regions in the yellow cluster should be excluded for variant calling. In other words, any alteration found in regions located in the yellow cluster should be filtered.</p>
<div class="figure align-center" id="id7">
<a class="reference internal image-reference" href="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/clusterplot.png"><img alt="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/clusterplot.png" src="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/clusterplot.png" style="width: 700px; height: 450px;" /></a>
<p class="caption"><span class="caption-text">Figure 5: <strong>Clustering of data from 1139 fastq files.</strong> Sequences of 5 bases (pentamer) are measured for quality score and reported in high or low score. Here, the percentage of the number a sequence is reported in high and low score is visualized. Following, clustering was conducted using K-means, thereby producing three clusters.</span></p>
</div>
<p>Together, fastq data indicates that quality score is important for identifying regions which are promising for mutation calling and which regions should be avoided. As described earlier, regions with a high quality score should be avoided while looking for mutations. In contrast, low quality score regions have the potential for identifying real mutations in ctDNA. For the construction of a data filter, regions should be selected from the clustering data. Nevertheless, only fastq files of one run on chromosome 9 have now been investigated and while this can provide for data filtering of this specific run, for data filter construction on a larger scale, multiple runs should be compared. Importantly, the quality scores of sequences can differ on the method being used. In this case, our method involves rolling circle amplification and nanopore sequencing of cfDNA. In order to make a sensitive data filter, a big database of healthy cfDNA should be investigated on quality score for sequences. In this way, the filter can exclude and include regions with respectively high and low quality scores. Furthermore, quality scores of sequences can differ on every run, causing some sequences to have higher or lower scores. This discrepancy should also be accounted for in the data filter. A possibility is to either include healthy cfDNA into every run or compare backbone sequences to identify run specific sequence quality score differences. Backbone analysis for identifying run specific changes in quality score sound more promising as the backbone is already incorporated in sequencing and the backbone sequence is known.</p>
<p><strong>p53 wild-type and mutant dataset analysis</strong>. So far, only run specific sequence quality scores have been discussed. However, our goal is to identify method specific high and low quality score regions. In order to visualize high and low score sequences specific for our method, data analysis should be conducted on multiple runs. Analysis of multiple runs can be simultaneously conducted using the High-Performance Computing (HPC) facility in the UMC, which will be discussed in the following segment. For these multiple datasets, aims are similar to the chromosome 9 dataset. Firstly, the mean should give an indication on whether changes in quality score occur in these datasets. Secondly, score plotting could tell us if regions exist. Finally, clustering would provide us with groups of regions to include and exclude in the data filter construction.</p>
<p>On the HPC, multiple ctDNA datasets derived from cyclomics are available for analysis, as they have been earlier constructed by the Kloosterman group. Here, the focus goes towards the rolling circle amplification of p53 mutated and wild-type(WT) datasets. The p53 gene codes for a tumor suppressor protein that can initiate apoptosis, arrest in the cell cycle and activate DNA repair proteins. In order to analyze these p53 datasets, a <a class="reference external" href="https://github.com/DouweSpaanderman/NaDA/blob/master/Scripts/fastq_qualityscore_analyser.py">fastq_script</a> was written to achieve similar data analysis and visualization as was applied to fastq files from chromosome 9. To improve data visibility, minor visualization updates were conducted.
Both mutated and WT datasets were separated in equally sized chunked files, around 4000 files each, and analysed as individual chunks to increase script parallelization, thus increasing speed. First, the datasets were investigated for changes in quality score, to give an indication whether these changes also harnessed in these datasets similarly to the chromosome 9 dataset. Therefore, the mean scores of all files were calculated and visualized in Figure 6.</p>
<div class="figure align-center" id="id8">
<a class="reference internal image-reference" href="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/RCAxMUT_WT_boxplot.png"><img alt="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/RCAxMUT_WT_boxplot.png" src="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/RCAxMUT_WT_boxplot.png" style="width: 700px; height: 450px;" /></a>
<p class="caption"><span class="caption-text">Figure 6: <strong>Boxplot of mean score from p53 mutated and WT dataset.</strong> For both datasets chunk 0 to 9 have been visualized. Chunk 10 to 24 were excluded, but showed similar results.</span></p>
</div>
<p>This boxplot clearly visualizes the lack of consistency between quality scores in the same sequence run. Similarly to our previous chromosome 9 dataset, Figure 6 visualizes the occurrence of difference in quality score in p53 datasets. However, the inconsistency between the different chunks in the same dataset reveals that quality scores are highly fluctuating between chunks. This could indicate that regions reported in high and low quality score also shift heavily between chunks.</p>
<p>While mean scores give an indication on quality score analysis, both quality score plotting and clustering are yet to determine whether high and low score regions exist and persist in multiple chunks and datasets. For every chunk derived from a dataset, sequences have been analyzed and visualized as tetramer, pentamer and hexamer. This was conducted to analyze the impact of different sized regions on the occurrence of high and low score region. Here, tetramers of wild-type chunk 0 to 3 have been plotted as shown in Figure 7.</p>
<div class="figure align-center" id="id9">
<a class="reference internal image-reference" href="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Fastq_gridplot_WT.png"><img alt="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Fastq_gridplot_WT.png" src="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Fastq_gridplot_WT.png" style="width: 700px; height: 450px;" /></a>
<p class="caption"><span class="caption-text">Figure 7: <strong>Tetramer sequence analysis for chunks zero to three of the p53 wild-type database.</strong> This figure illustrates the mean quality score for each tetramer possible in one chunk. A) chunk 0. B) chunk 1. C) chunk 2. D) chunk 3. The interactive figure can be found here and also visualizes data analysis when divided into high, medium and low group.(<a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/source/_static/RCAxWT_chunk0_4.0_3.0_score_plotting.html">WT_chunk0</a>, <a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/source/_static/RCAxWT_chunk1_4.0_3.0_score_plotting.html">WT_chunk1</a>, <a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/source/_static/RCAxWT_chunk2_4.0_3.0_score_plotting.html">WT_chunk2</a> and <a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/source/_static/RCAxWT_chunk3_4.0_3.0_score_plotting.html">WT_chunk3</a>)</span></p>
</div>
<p>Similarly to the boxplot, there seems to be a lack of consistency between chunks as high reported tetramers differ heavily between these chunks. Similar results are visible for bigger sized chunks(<a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/build/html/Supplementary.html">supplementary_1</a>) and chunks derived from the p53 mutant dataset(<a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/build/html/Supplementary.html">supplementary_2</a>). These datasets show that there is yet to be proven that a correlation between quality scores and specific regions or chunks exists. However, clustering could clarify the occurrence of high quality score regions by better identification of these regions. In order to cluster data derived from dataset chunks, the same algorithm (K-Means) is used as described above. In Figure 8 clustering of chunks, 0 to 3 from p53 WT has been visualized.</p>
<div class="figure align-center" id="id10">
<a class="reference internal image-reference" href="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Fastq_gridplot_WT_cluster.png"><img alt="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Fastq_gridplot_WT_cluster.png" src="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Fastq_gridplot_WT_cluster.png" style="width: 700px; height: 450px;" /></a>
<p class="caption"><span class="caption-text">Figure 8: <strong>Clustering of hexameer sequence for chunks zero to three of the p53 wild-type database.</strong> Data points are visualised as percentage reported in high (y-axis) and low score(x-axis). A) chunk 0. B) chunk 1. C) chunk 2. D) chunk 3. Interactive cluster plot can be found here. (<a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/source/_static/RCAxWT_chunk0_6.0_5.0_score_clustering.html">WT_chunk0_cluster</a>, <a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/source/_static/RCAxWT_chunk1_6.0_5.0_score_clustering.html">WT_chunk1_cluster</a>, <a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/source/_static/RCAxWT_chunk2_6.0_5.0_score_clustering.html">WT_chunk2_cluster</a> and <a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/source/_static/RCAxWT_chunk3_6.0_5.0_score_clustering.html">WT_chunk3_cluster</a>)</span></p>
</div>
<p>By clustering, regions can be identified that have both been reported often as high score and few times as low scores. However, between chunks there is a huge discrepancy in quality scores. Chunk 0 and 2 have an overall much lower quality score in comparison with chunk 1 and 2, which was also identified with mean quality scores in Figure 6. This big difference in overall quality of the chunks suggests that quality score cannot be used for constructing a data filter as regions vary too much. Nevertheless, if regions are always present in the same cluster between chunks, high and low quality score regions could still be identified. Additionally, quality score means should not have to influence score from a single region in comparison to other regions in the same chunk. With this in mind, future analysis should be conducted comparing regions between chunks for the same group. For example, regions reported in the high score group from chunk 0 should be compared to regions reported in the high score groups of the other chunks.</p>
<p>As described earlier, identifying high quality score regions could help data filtering for mutations, as bases in regions with high quality score are less likely to be rightfully assigned and the other way around for low quality score region. Therefore, respectively, these regions could be excluded or included in variant calling. Currently, quality score analysis shows a lot of inconsistency between chunks of the same dataset. Thus, it is yet impossible to conclude any regions that have either a high or low quality score. Therefore, quality score has currently no application in creating a data filter. Altogether, although quality score shows limited possibilities, regions should still be compared between chunks, because regions could still be reported in similar clusters (high score or low score clusters). Further comparison of regions between chunks could clarify whether there is a correlation between regions and quality scores.</p>
</div>
<div class="section" id="variant-call-format-files">
<h2>Variant Call Format files<a class="headerlink" href="#variant-call-format-files" title="Permalink to this headline">¶</a></h2>
<p><strong>Mutation distribution of single nucleotide polymorphisms</strong>. After base calling is conducted, reads are compared with the reference genome in order to call variants occurring in the samples. The output of variant calling is located in Variant Call Format (VCF) files, which are in text-based formats similar to fastq, containing data of single positions bases in the genome. In these files, variants are formatted with the reference included. For sequenced sites, the number of reads found with the mutation and reference is given.</p>
<p>In these VCF files multiple variances can be identified depending on which dataset is used. These variances identified in the VCF files can be real mutations but also PCR and sequencing errors. Analysing VCF files could yield information about real mutations positions as well as the positions where errors are more prone to be located. Similarly to fastq files, it could be possible that errors occur more often in specific regions of bases. Therefore, Alterations found in these locations could be excluded by a data filter. Our goals for VCF analysis are to be able to identify real driver mutations and to analyze if regions can be identified which are more prone to errors. Identifying real driver mutations can be done by comparing wild-type (WT) and mutated (Mut) databases as driver mutations should not be included in the WT. Furthermore, region analysis can be conducted by constructing larger sequences similar to fastq analysis.</p>
<p>In order to achieve our goals, a dataset was analyzed derived from the Cyclomics project, in which sequencing was performed with Nanopore. The data contains a part of the p53 gene on chromosome 17, existing of 160 nucleotides and a backbone, which was used for the circular PCR reaction. In total, 1187 VCF files were used for variant calling. These VCF files were screened for single nucleotide polymorphism (SNP) occurrence. Files were stripped of reported mutated bases and other data was discarded. As described earlier, every variant site has a number of reads that cover this site. These reads can be both coupled to the mutation and the reference. For example, on position 7577503 a SNP was found in 6 reads and 3 reads were coupled to the reference. While normally the number of reads coupled to the mutation in contrast to the total amount of reads is important, here the occurrence of certain SNPs has been firstly investigated. Analysing SNPs in ctDNA could help identify real driver mutations present in a tumor. Likewise, analyzing cfDNA for SNPs by identifying passenger mutation, asymmetric DNA errors, PCR errors and sequencing errors can be useful for lowering background noise.</p>
<p>Our first aim is to analyze and visualize SNP occurrences in the p53 databases. Identifying often manifesting alteration can help understand what bases are more prone to errors. Normally, no discrepancy between alterations should occur as biological mechanisms do not favor specific base substitutions. Therefore, changes in Cyclomics can be devised to suppress these errors. In order to investigate the amount of SNPs in the files, VCF files were similarly stripped as Fastq files and separated by either sequence or backbone. Next, for the alterations, a parameter was set at a minimum of 25 percent of the reads that should be coupled to the mutant variant and visualized in Figure 9:</p>
<div class="figure align-center" id="id11">
<a class="reference internal image-reference" href="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Combined_vcf_snp_analysis.png"><img alt="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Combined_vcf_snp_analysis.png" src="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Combined_vcf_snp_analysis.png" style="width: 700px; height: 450px;" /></a>
<p class="caption"><span class="caption-text">Figure 9: <strong>Distribution of SNPs in the sequence of the p53 gene for 1187 VCF files.</strong> Parameter for variant identification was set at 25% of the reads to the variant. Variants are displayed as C &gt; T, meaning that T substitutes C. A) Bar plot with single nucleotide polymorphisms occurrence as percentage of whole. B) Heatmap of the same variances with the number of occurrences in the files</span></p>
</div>
<p>Both Figure 9A and 9B illustrate a common occurrence of G &gt; A mutation and to lesser extend C &gt; A. The high prevalence of these SNPs in contrast to other alterations are a strong indication that these alterations are caused by a non-biological mechanism, which can be errors in the rolling circle amplification, library preparation and sequencing of the ctDNA. In literature, cytosine deamination has been described to increase C:G &gt; T:A noise levels (<a class="reference external" href="http://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/build/html/References.html">8</a>). Cytosine deamination occurs when the amino group of cytosine is removed resulting in a change to a uracil analog. Also, the less occurring alteration C &gt; A has been reported to be caused by oxidative DNA damage during sample preparation(<a class="reference external" href="http://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/build/html/References.html">9</a>). Both these types of alterations can be a result of polymerase-induced errors. Possible suggested methods to suppress these errors are adding DNA repair mechanisms upon polymerase chain reaction (PCR) and lowering heat. However, if results are similar in other databases, an in silico approach to polish background noise can also be devised. Overall, optimizing Cyclomics to minimize error occurrences should first be conducted.</p>
<p><strong>Region SNP analysis</strong>. As described earlier, one of our goals is to identify if regions are present in our sample which are more prone to errors. Therefore, trimers and pentamers were selected in which the middle base was reported to contain a SNPs in some of the files for heatmap analysis. Creating chunks for SNP analysis instead of single base analysis was conducted to visualize sequences that were more likely to contain SNPs. Pandas library was used to create a dataframe for the number of times mutation occurred to either A, T, C or G. This dataframe was then mapped to a heatmap with the reference sequence (Figure 10). Just as in previous Figures, length of the surrounding bases can be changed to give a wider variety of information. Firstly, analyzing of this p53 dataset visualized the occurrence of regions that are prone to errors. Secondly, This Figure gave more information about base combinations with high alteration affinity, such as CGC to CAC.</p>
<div class="figure align-center" id="id12">
<a class="reference internal image-reference" href="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Variance_occurence_in_sequence_vcf_3.PNG"><img alt="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Variance_occurence_in_sequence_vcf_3.PNG" src="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Variance_occurence_in_sequence_vcf_3.PNG" style="width: 700px; height: 450px;" /></a>
<p class="caption"><span class="caption-text">Figure 10: <strong>Occurence of variance per reference sequence to different bases.</strong> In all the sequences the middle base is reported to be mutated in some of the vcf files. This mutation again has a parameter that is set at 25% of the reads at least mutated.</span></p>
</div>
<p>Identifying high variance chunks in healthy cfDNA is important to understand where errors arise in the sequence and thus could help polishing background noise by deselecting these chunks for variant calling analysis. However, it is important to keep in mind that variance present in WT samples could not only be errors but also real passenger mutations.</p>
<p>Furthermore, just as with the fastq files, variances can be separated between alterations specific for a run and alterations specific for the method being used. For instance, CTC &gt; A could be an alteration that is specifically highly mutated in a particular run, while CGC &gt; A occurs often in every run with this method of rolling circle amplification and Nanopore sequencing. Therefore, filtering should be able to account for both run specific and method specific alterations. In the same manner, a large database of healthy cfDNA could accomplish a method specific filter and adding healthy cfDNA into every run could provide with a specific alteration filter, which will be helpful in lowering background noise. Also more convenient, backbone data could be used to identify run specific errors as the backbone does not change between runs and should thus never contain alterations.</p>
<p><strong>p53 wild-type and mutant dataset analysis</strong>. In the previous segment, only one run was investigated for the presence of SNPs. In order to establish enough information for specifying a method specific data filter, multiple Cyclomics databases should be analyzed for occurring alterations. In the following segments analysis of four datasets was conducted. These datasets all cover the p53 gene, either WT, Mut, 1% mutated or 10% mutated. Investigation of these datasets could provide for real driver mutations and often error harnessing regions. Further analysis on these VCF files was conducted on the HPC system as it required greater computer performance and parallelization. In order to conduct VCF analysis on the HPC, a VCF analysis <a class="reference external" href="https://github.com/DouweSpaanderman/NaDA/blob/master/Scripts/vcf_snp_variant_analyser.py">script</a> was written. This script analyses mutation occurrences in the insert and backbone similarly as described above. Subsequently, visualization was updated, presenting sequences as a percentage of times it has been reported to contain a SNP to how many times it was reported in total for the whole dataset.</p>
<p>Firstly, SNPs were investigated for their individual positions. This was conducted to be able to identify the real driver mutation by comparing wild-type and mutated datasets. As visualized in Figure 11, using this single position heatmap, a clear mutation, position 7578265 A &gt; T, can be found in the p53 mutated dataset.</p>
<div class="figure align-center" id="id13">
<a class="reference internal image-reference" href="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/single_base_insert_MUT.PNG"><img alt="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/single_base_insert_MUT.PNG" src="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/single_base_insert_MUT.PNG" style="width: 700px; height: 450px;" /></a>
<p class="caption"><span class="caption-text">Figure 11: <strong>Single position SNPs analysis for p53 mutated database</strong>. Occurrence of SNP is visualized as a percentage of times it has been reported with a SNP to the total amount of times it has been reported in the dataset. Interactive figure can be found here(<a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/source/_static/RCAxMUT_single_base_insert_1_heatmap_sequences.html">single_base_insert_MUT</a>)</span></p>
</div>
<p>Similar to Figure 11, wild-type, 1% and 10% have been analyzed for position specific SNPs (<a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/build/html/Supplementary.html">supplementary_3</a>). As expected, wild-type shows no occurrence of the specific mutation and the other datasets are in concordance with the percentages of reads that should contain the SNP. Therefore, this <a class="reference external" href="https://github.com/DouweSpaanderman/NaDA/blob/master/Scripts/vcf_snp_variant_analyser.py">script</a> is able to identify real driver mutations in this p53 dataset, which is A to T at position 7578265.</p>
<p>However, while the expected alteration is found in our datasets, for those datasets other alteration were also identified. These alterations were not in line with our expectations and could have a number of explanations. Firstly, it could be a real passenger mutation which is not normally present in the reference. Secondly, these alterations could be caused by protocol errors, such as PCR errors and sequencing errors. For both these types of alterations could be accounted for by respectively excluding certain positions and certain regions of bases in the data filter. However, the distinction between these alterations is difficult to determine. Currently, found alterations can help to construct a data filter specific for p53 databases as specific positions can be excluded. Nevertheless, our aim is to construct a data filter that can be applied to multiple gene databases. Therefore, regions should further be investigated to determine if they are reported as mutated because of errors or real passenger mutations.</p>
<p>In order to construct a data filter usable for multiple genes, regions rather than single positions were investigated for SNPs. Region selection could help identifying SNPs due to protocol errors and help exclude them from the analysis. First, chunks of trimers and pentamers were selected as described in region SNP analysis. Subsequently, these chunks were plotted in a heatmap visualized in Figure 12. Mutation occurrence (GGATA &gt; T) could again be clearly visualized in this dataset. Furthermore, WT analysis showed a complete absence of this mutation (<a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/build/html/Supplementary.html">supplementary_4</a>) and the other two datasets are in concordance with the percentage mutated (<a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/build/html/Supplementary.html">supplementary_5</a>).</p>
<p>Extraordinarily, apart from the expected mutation, in all four datasets other pentamers seem to be mutated as well. Especially, CAACC is reported to be highly mutated (around 30%) for all the datasets. This could indicate either the occurrence of other mutations in the dataset or the identification of pentamers which cause errors throughout our workflow.</p>
<div class="figure align-center" id="id14">
<a class="reference internal image-reference" href="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Variance_occurence_in_MUT.PNG"><img alt="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Variance_occurence_in_MUT.PNG" src="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Variance_occurence_in_MUT.PNG" style="width: 700px; height: 450px;" /></a>
<p class="caption"><span class="caption-text">Figure 12: <strong>SNPs analysis for pentameer chunks</strong>. Occurrence of SNP is visualized as a percentage of the amount of time sequence has been reported in the dataset. In all the chunks the middle base has been reported to be mutated in some of the vcf files. The dataset used here is the p53 mutated dataset. The interactive figure can be found here(<a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/source/_static/RCAxMUT_insert_5.0_heatmap_sequences.html">MUT_heatmap</a>).</span></p>
</div>
<p>Importantly, the identified alterations could be present anywhere the pentamer is located in the sequence. Furthermore, if an alteration occurs in one specific position it is more prone to be an actual mutation. In contrast, alterations occurring in multiple locations within the same pentamer could indicate a systematic problem with PCR or sequencing causing a falsely identified mutation. In order to visualize the location of the altered chunks, another script was prepared to compare the pentamer locations in the sequence. This script constructed an <a class="reference external" href="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Excel.xlsx">excel</a> with the position of the chunks, including SNP, the percentage of times this particular chunk with SNP occurs in a certain location and the number of times this specific alteration was found in the dataset.</p>
<p>As expected, the p53 mutated altered chunk GGATA &gt; T on locations 7578263 to 7578267 was reported as the only occurrence of this particular sequence with alteration in the whole dataset. This finding empowers the conclusion that this is, in fact, a real mutation. Similarly, other found alterations such as CAACC &gt; G at position 7578333 to 7578337 were also only reported here in the whole sequence, suggesting that this is also a real mutation. However, this alteration has not been reported yet in the GRCh37.13 reference genome(<a class="reference external" href="http://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/build/html/References.html">10</a>), questioning the importance of this mutation. Also, this alteration is located in an intron and is thus less likely to be a driver mutation. Altogether, a mutation like CAACC &gt; G is most likely a real passenger mutation and this location should thus be excluded by our data filter. Besides this passenger mutation, our datasets also show signs of error-prone regions. In figure 12, alteration CTGGG &gt; A was found with an occurrence of 7,5%. Importantly, location analysis described in <a class="reference external" href="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Excel.xlsx">excel</a> showed the presence of this alteration in multiple locations as both position 7578321 and 7578330 harnessed this specific alteration. Therefore, CTGGG &gt; A is a region that is most likely more prone to errors occurring in our method and should be excluded by our data filter.</p>
<p>Future investigations should be conducted towards automation of this process of identifying reported alterations by error-prone regions or real passenger mutations. Furthermore, automation of this process should also check if the mutation is reported in the GRCh37.13 reference genome(<a class="reference external" href="http://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/build/html/References.html">10</a>). Automation of this process is important for further investigation of VCF files for data filter construction.</p>
<p>Currently, we are able to identify SNPs occurring in ctDNA when we compare them to cfDNA. Additionally, we are able to visualize all SNPs occurring in a dataset and are already able to use this knowledge to analyze p53 datasets. Also, error-prone regions have been shown to persist in our database and can be further investigated for data filter construction. As a future perspective identification of both passenger mutation and error-prone regions should be automated to make data filter construction easier.</p>
<p>Apart from insert data, the backbone was also analyzed for SNPs. Backbone data could be applied in identifying run specific alterations, which are errors that occur in the sequence and differ in each run. After identification, these run specific errors could be included in data filtering. Firstly, backbone data was similarly stripped and analyzed as insert data and also visualized using heatmap. In Figure 13 backbone heatmaps from p53 wild-type and 10% mutated were visualized, which clearly shows the occurrence of run specific errors.</p>
<div class="figure align-center" id="id15">
<a class="reference internal image-reference" href="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/backbone_variance_WT_10.png"><img alt="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/backbone_variance_WT_10.png" src="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/backbone_variance_WT_10.png" style="width: 700px; height: 450px;" /></a>
<p class="caption"><span class="caption-text">Figure 13: <strong>SNPs analysis for backbone</strong>. Occurrence of SNP is visualized as a percentage of the amount of time sequence has been reported in backbone. In all the chunks the middle base has been reported to be mutated in some of the vcf files. A) Backbone from p53 wild-type. B) Backbone from p53 10% mutated. Interactive figure can be found here(<a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/source/_static/RCAxWT_backbone_5.0_heatmap_sequences.html">Backbone_WT</a> and <a class="reference external" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/source/_static/RCAxPool5_xI2_backbone_5.0_heatmap_sequences.html">Backbone_10</a>).</span></p>
</div>
<p>Similar to the insert data, the backbone data has been investigated for SNP locations (<a class="reference external" href="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Excel.xlsx">excel</a>). Because the backbone should not contain any SNPs, applying reported SNPs to the data filter could provide for run specific background noise canceling. Importantly, SNPs should at least have a 10% occurrence in the backbone to be applied in the data filter, to limit sequence exclusion. Furthermore, automation should be conducted to analyze error-prone regions.</p>
</div>
<div class="section" id="script-tests">
<h2>Script Tests<a class="headerlink" href="#script-tests" title="Permalink to this headline">¶</a></h2>
<p>Before functions and scripts are run over multiple files and directories, they should be checked for quality. In order to check a function for its functionality, test scripts can be written. These testing scripts use the assert function to identify whether the set criteria are met. As an example the earlier described parse_fasta_file_error function is checked for its quality with the following testing script:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">TestDoneFastqParser</span><span class="p">:</span>

        <span class="k">def</span> <span class="nf">setup_method</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
                <span class="n">sequence_file</span> <span class="o">=</span> <span class="s1">&#39;C:/Users/Douwe/Documents/Python/test_cases/test_fastq2.done_fastq&#39;</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">dl</span><span class="o">.</span><span class="n">parse_fasta_file_error</span><span class="p">(</span><span class="n">sequence_file</span><span class="p">)</span>
                <span class="n">id_</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">id_</span><span class="p">][</span><span class="s1">&#39;score&#39;</span><span class="p">]</span>

        <span class="k">def</span> <span class="nf">check_valid_DNA_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">upper</span><span class="p">()):</span>
                        <span class="k">if</span> <span class="ow">not</span> <span class="n">l</span> <span class="ow">in</span> <span class="s1">&#39;ACTGN&#39;</span><span class="p">:</span>
                                <span class="k">return</span> <span class="kc">False</span>
                <span class="k">return</span> <span class="kc">True</span>

        <span class="k">def</span> <span class="nf">test_has_id</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
                <span class="k">for</span> <span class="nb">id</span> <span class="ow">in</span> <span class="s1">&#39;@&#39;</span><span class="p">:</span>
                        <span class="k">assert</span> <span class="nb">id</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">]</span>

        <span class="k">def</span> <span class="nf">test_sequence_correct</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                        <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_valid_DNA_sequence</span><span class="p">(</span><span class="n">v</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">])</span> <span class="o">==</span> <span class="kc">True</span>

        <span class="k">def</span> <span class="nf">test_score_correct</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">letter</span> <span class="ow">in</span> <span class="s1">&#39;ABCDEFGHIJKLMNOPQRSTUVWXYZ&#39;</span><span class="p">:</span>
                        <span class="k">assert</span> <span class="n">letter</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">score</span>
</pre></div>
</div>
<p>The class function is used to define which function is going to be checked for quality. Firstly, the function is setup with a test file, which is designed to identify flaws in the function. In other words, the test file consists of a lot of errors which the function should not pickup. Next, multiple assertions are made, such as the assertion that letters in sequence can only consist of A, C, T, G and N. Also, the score should consist of characters and not involve any letters. While this is an example of a test script, every function and script is checked for their quality. Testing scripts can be found <a class="reference external" href="https://github.com/DouweSpaanderman/NaDA/tree/master/Testing%20scripts">here</a>.</p>
<div class="line-block">
<div class="line"><br /></div>
<div class="line"><br /></div>
</div>
<div class="figure align-center">
<a class="reference external image-reference" href="https://htmlpreview.github.io/?https://github.com/DouweSpaanderman/NaDA/blob/master/Documentation/build/html/Conclusion.html"><img alt="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Next.png" src="https://raw.githubusercontent.com/DouweSpaanderman/NaDA/master/Documentation/source/_static/Next.png" style="width: 100px; height: 100px;" /></a>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">Internship report</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="General concept.html">General concept</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Results and discussion</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#fastq-files">Fastq files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#variant-call-format-files">Variant Call Format files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#script-tests">Script Tests</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Conclusion.html">Conclusion</a></li>
<li class="toctree-l1"><a class="reference internal" href="References.html">References</a></li>
<li class="toctree-l1"><a class="reference internal" href="Supplementary.html">Supplementary</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="General concept.html" title="previous chapter">General concept</a></li>
      <li>Next: <a href="Conclusion.html" title="next chapter">Conclusion</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, Douwe Spaanderman.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/Results and Discussion.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>