
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Supplementary &#8212; Internship report 3.0a1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '3.0a1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="References" href="References.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="supplementary">
<h1>Supplementary<a class="headerlink" href="#supplementary" title="Permalink to this headline">Â¶</a></h1>
<div class="figure align-center" id="id2">
<a class="reference internal image-reference" href="_images/Fastq_gridplot_WT_cat.png"><img alt="_images/Fastq_gridplot_WT_cat.png" src="_images/Fastq_gridplot_WT_cat.png" style="width: 1079.3999999999999px; height: 443.79999999999995px;" /></a>
<p class="caption"><span class="caption-text">Supplementary Figure 1: <strong>Hexameer sequence analysis devided in category high, medium and low for chunk 0 and 1 of the p53 wild-type database.</strong> Instead of mean qualityscore, data analysis is manipulated by assigning Hexameer, when found, into either high, medium or low score and visualized here. Interactive figure as well as mean qualityscore plot of these chunks can be found here. (<a class="reference external" href="C:\Users\Douwe\Documents\GitHub\NaDA\Documentation\source\_static\RCAxWT_chunk0_6.0_5.0_score_plotting.html">WT_chunk0</a>, <a class="reference external" href="C:\Users\Douwe\Documents\GitHub\NaDA\Documentation\source\_static\RCAxWT_chunk1_6.0_5.0_score_plotting.html">WT_chunk1</a>)</span></p>
</div>
<div class="figure align-center" id="id3">
<a class="reference internal image-reference" href="_images/Fastq_gridplot_MUT.png"><img alt="_images/Fastq_gridplot_MUT.png" src="_images/Fastq_gridplot_MUT.png" style="width: 1062.6px; height: 890.4px;" /></a>
<p class="caption"><span class="caption-text">Supplementary figure 2: <strong>pentameer sequence analysis for chunks 0-2 and 4 of the p53 mutant database.</strong> Figure illustrates the mean qualityscore for each tetrameer possible in one chunk. The interactive figure can be found here and visualizes data analysis when devided into high, medium and low group. (<a class="reference external" href="C:\Users\Douwe\Documents\GitHub\NaDA\Documentation\source\_static\RCAxMUT_chunk0_4.0_3.0_score_plotting.html">MUT_chunk0</a>, <a class="reference external" href="C:\Users\Douwe\Documents\GitHub\NaDA\Documentation\source\_static\RCAxMUT_chunk1_4.0_3.0_score_plotting.html">MUT_chunk1</a>, <a class="reference external" href="C:\Users\Douwe\Documents\GitHub\NaDA\Documentation\source\_static\RCAxMUT_chunk2_4.0_3.0_score_plotting.html">MUT_chunk2</a> and <a class="reference external" href="C:\Users\Douwe\Documents\GitHub\NaDA\Documentation\source\_static\RCAxMUT_chunk4_4.0_3.0_score_plotting.html">MUT_chunk4</a>)</span></p>
</div>
<div class="figure align-center" id="id4">
<a class="reference internal image-reference" href="_images/Variance_occurence_in_WT.png"><img alt="_images/Variance_occurence_in_WT.png" src="_images/Variance_occurence_in_WT.png" style="width: 791.0px; height: 684.5999999999999px;" /></a>
<p class="caption"><span class="caption-text">Figure a: A) Occurence of alterations as a percentages of the amount of times sequence has been reported in the dataset. In all the sequences the middle base has been reported to be mutated in some of the vcf files. B) plot of the distribution of SNPs. Dataset used here is p53 wild-type, without a mutation on position 7578265 in chromosome 17. Interactive figure can be found here(<a class="reference external" href="C:\Users\Douwe\Documents\GitHub\NaDA\Documentation\source\_static\RCAxWT_insert_5.0_heatmap_sequences">WT_heatmap</a> and <a class="reference external" href="C:\Users\Douwe\Documents\GitHub\NaDA\Documentation\source\_static\RCAxWT_insert_5.0_SNP_plot">WT_snp_plot</a>)</span></p>
</div>
<div class="figure align-center" id="id5">
<a class="reference internal image-reference" href="_images/Variance_occurence_in_MUT_1%_10%.png"><img alt="_images/Variance_occurence_in_MUT_1%_10%.png" src="_images/Variance_occurence_in_MUT_1%_10%.png" style="width: 827.4px; height: 663.5999999999999px;" /></a>
<p class="caption"><span class="caption-text">Figure b: A) Occurence of alterations as a percentages of the amount of times sequence has been reported in the dataset. In all the sequences the middle base has been reported to be mutated in some of the vcf files. Dataset used here is 1% p53 mutant. B) Dataset used here is 10% p53 mutant. Interactive figure can be found here(<a class="reference external" href="C:\Users\Douwe\Documents\GitHub\NaDA\Documentation\source\_static\RCAxPool5_xI1_insert_5.0_heatmap_sequences.html">1_heatmap</a> and <a class="reference external" href="C:\Users\Douwe\Documents\GitHub\NaDA\Documentation\source\_static\RCAxPool5_xI2_insert_5.0_heatmap_sequences.html">10_heatmap</a>)</span></p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">Internship report</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Table of Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="General workflow.html">General workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="Results and Discussion.html">Results and discussion</a></li>
<li class="toctree-l1"><a class="reference internal" href="References.html">References</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Supplementary</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="References.html" title="previous chapter">References</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, Douwe Spaanderman.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/Supplementary.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>